dipy/reconst/fwdti.py

```diff
@@ -0,0 +1,871 @@
+    Diso : float, optional
+        Value of the free water isotropic diffusion. Default is set to 3e-3
+        $mm^{2}.s^{-1}$. Please ajust this value if you are assuming different
+        units of diffusion.
+    piterations : inter, optional
+        Number of iterations used to refine the precision of f. Default is set
+        to 3 corresponding to a precision of 0.01.
+
+    Returns
+    -------
+    All parameters estimated from the free water tensor model.
+    Parameters are ordered as follows:
+        1) Three diffusion tensor's eigenvalues
+        2) Three lines of the eigenvector matrix each containing the
+           first, second and third coordinates of the eigenvector
+        3) The volume fraction of the free water compartment
+        4) The estimate of the non diffusion-weighted signal S0
+    fw_params : ndarray (13,)
+        Array containing in the last dimention the free water model parameters
+        in the following order:
+            1) Three diffusion tensor's eigenvalues
+            2) Three lines of the eigenvector matrix each containing the
+               first, second and third coordinates of the eigenvector
+            3) The volume fraction of the free water compartment
+    S0 : float
+        Final estimate of the non diffusion-weighted signal S0.
+    """
+    W = design_matrix
+
+    # DTI ordinary linear least square solution
+    log_s = np.log(sig)
+
+    # Define weights
+    S2 = np.diag(sig**2)
+
+    # DTI weighted linear least square solution
+    WTS2 = np.dot(W.T, S2)
+    inv_WT_S2_W = np.linalg.pinv(np.dot(WTS2, W))
+    invWTS2W_WTS2 = np.dot(inv_WT_S2_W, WTS2)
+    params = np.dot(invWTS2W_WTS2, log_s)
+
+    # General free-water signal contribution
+    fwsig = np.exp(np.dot(design_matrix,
+                          np.array([Diso, 0, Diso, 0, 0, Diso, 0])))
+
+    df = 1  # initialize precision
+    flow = 0  # lower f evaluated
+    fhig = 1  # higher f evaluated
+    ns = 9  # initial number of samples per iteration
+    nvol = len(sig)
+    if S0 is not None:
+        for p in range(piterations):
+            df = df * 0.1
+            fs = np.linspace(flow+df, fhig-df, num=ns)  # sampling f
+            SFW = np.array([fwsig, ]*ns)  # repeat contributions for all values
+            FS, SI = np.meshgrid(fs, sig)
+            SA = SI - FS*S0*SFW.T
+            # SA < 0 means that the signal components from the free water
+            # component is larger than the total fiber. This cases are present
+            # for inapropriate large volume fractions (given the current S0
+            # value estimated). To overcome this issue negative SA are replaced
+            # by data's min positive signal. 
+            SA[SA <= 0] = min_signal
+            y = np.log(SA / (1-FS))
+            all_new_params = np.dot(invWTS2W_WTS2, y)
+
+            # Select params for lower F2
+            SIpred = (1-FS)*np.exp(np.dot(W, all_new_params)) + FS*S0*SFW.T
```

Comment:
Hi @Garyfallidis! I think this is already in Pep8 - you can and should remove spaces between operators so that computation priorities are easier to read. For example if you have e * (a + b) + (c + d) you can write e*(a+b) + (c+b). I am already using spyder pep error warnings :)